drivers/adc/ad7616/iio_ad7616.c

```diff
@@ -0,0 +1,295 @@
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	int ret;
+	int val;
+	enum ad7616_range range;
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	struct ad7616_dev *dev = device;
+
+	ret = sscanf(buf, "%u", &val);
+	if (ret != 1)
+		return -EINVAL;
+
+	switch (val) {
+	case 10:
+		range = AD7616_10V;
+		break;
+	case 5:
+		range = AD7616_5V;
+		break;
+	case 2:
+		range = AD7616_2V5;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ad7616_set_range(dev, ch, range);
+}
+
+/**
+ * @brief	Read buffer data corresponding to AD7616 IIO device
+ * @param	iio_dev_data[in] - Pointer to IIO device data structure
+ * @return	0 in case of success, negative error code otherwise
+ */
+static int32_t iio_ad7616_submit_buffer(struct iio_device_data *iio_dev_data)
+{
+	uint16_t ret;
+	struct ad7616_conversion_result results[8];
+	uint32_t bytes_per_scan = iio_dev_data->buffer->bytes_per_scan;
+	uint16_t data[2];
+	uint32_t i = 0;
+
+	for (; i < iio_dev_data->buffer->samples; i++) {
+		ret = ad7616_read_data_serial(iio_dev_data->dev, results, 1);
+		if (ret)
+			return ret;
+
+		data[0] = results[0].channel_a;
+		data[1] = results[0].channel_b;
+
+		ret = no_os_cb_write(iio_dev_data->buffer->buf, data, bytes_per_scan);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+static struct scan_type ad7616_iio_scan_type = {
+	.sign = 's',
+	.realbits = CHN_STORAGE_BITS,
+	.storagebits = CHN_STORAGE_BITS,
+	.shift = 0,
+	.is_big_endian = false
+};
+
+static struct iio_attribute ad7616_iio_ch_attributes[] = {
+	{ .name = "raw", .show = get_raw, .store = set_raw },
+	{ .name = "scale", .show = get_scale, .store = set_scale },
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_attribute ad7616_iio_global_attributes[] = {
+	AD7616_CHN_ATTR("crc", CRC_ATTR_ID),
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_channel ad7616_channels[] = {
+	AD7616_CH(0),
+	AD7616_CH(1),
+	AD7616_CH(2),
+	AD7616_CH(3),
+	AD7616_CH(4),
+	AD7616_CH(5),
+	AD7616_CH(6),
+	AD7616_CH(7),
+	AD7616_CH(8),
+	AD7616_CH(9),
+	AD7616_CH(10),
+	AD7616_CH(11),
+	AD7616_CH(12),
+	AD7616_CH(13),
+	AD7616_CH(14),
+	AD7616_CH(15),
+};
+
+static struct iio_device ad7616_iio_device_template = {
+	.num_ch = NO_OS_ARRAY_SIZE(ad7616_channels),
+	.channels = ad7616_channels,
+	.attributes = ad7616_iio_global_attributes,
+	.debug_reg_read = iio_ad7616_debug_reg_read,
+	.debug_reg_write= iio_ad7616_debug_reg_write,
+	.submit = iio_ad7616_submit_buffer,
+};
+
+/**
+ * @brief	Init for reading/writing and parameterization of a
+ *			ad7616 IIO device
+ * @param	desc[in,out] - IIO device descriptor
+ * @return	0 in case of success, negative error code otherwise
+ */
+int32_t ad7616_iio_param_init(struct iio_device **desc)
+{
+	struct iio_device *iio_ad7616_inst;
+
+	iio_ad7616_inst = calloc(1, sizeof(struct iio_device));
+	if (!iio_ad7616_inst)
+		return -ENOMEM;
+
+	*iio_ad7616_inst = ad7616_iio_device_template;
```

Comment:
so in the newer drivers we have a different way of implementing iio drivers. the device driver is initialized here (and not in the example application). you may take a look at iio_adxl355 as an example

Reply:
I'm trying to use `iio_app_init` but at the moment I'm getting some random bytes out of the uart, even If I did not change the `uart_init_param`

Reply:
the default message will always be shown at 115200, after that the uart is reconfigured with the chosen baudrate (in this case 230400). make sure to use 230400 in IIO Oscilloscope connection.
the alignment is not an issue. it is simply a matter of remapping LF to CR + LF in your terminal emulator.